filters/auth/oauth.go

```diff
@@ -0,0 +1,429 @@
+	switch s {
+	case AuthAllScopeName:
+		return checkAllScopes
+	case AuthAnyScopeName:
+		return checkAnyScopes
+	case AuthAnyKVName:
+		return checkAnyKV
+	case AuthAllKVName:
+		return checkAllKV
+	}
+	return checkUnknown
+}
+
+func (s *spec) Name() string {
+	switch s.typ {
+	case checkAnyScopes:
+		return AuthAnyScopeName
+	case checkAllScopes:
+		return AuthAllScopeName
+	case checkAnyKV:
+		return AuthAnyKVName
+	case checkAllKV:
+		return AuthAllKVName
+	}
+	return AuthUnknown
+}
+
+// CreateFilter creates an auth filter. All arguments have to be
+// strings. The first argument is the realm and the rest the given
+// scopes to check. How scopes are checked is based on the type. The
+// shown example will grant access only to tokens from `myrealm`, that
+// have scopes `read-x` and `write-y`:
+//
+//     s.CreateFilter("myrealm", "read-x", "write-y")
+//
+func (s *spec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) == 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	if s.typ == checkUnknown {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	ac := &authClient{urlBase: s.cfg.Endpoint.TokenURL}
+
+	f := &filter{typ: s.typ, authClient: ac, kv: make(map[string]string)}
+	if len(sargs) > 0 {
+		switch f.typ {
+		case checkAnyKV:
+			fallthrough
+		case checkAllKV:
+			f.realm = sargs[0]
+			sargs = sargs[1:]
+			for i := 0; i+1 < len(sargs); i += 2 {
+				f.kv[sargs[i]] = sargs[i+1]
+			}
+			if len(sargs) == 0 || len(sargs)%2 != 0 {
+				return nil, filters.ErrInvalidFilterParameters
+			}
+		default:
+			f.realm, f.scopes = sargs[0], sargs[1:]
+		}
+
+	}
+
+	return f, nil
+
+}
+
+func (kv kv) String() string {
+	var res []string
+	for k, v := range kv {
+		res = append(res, k, v)
+	}
+	return strings.Join(res, ",")
+}
+
+// String prints nicely the filter configuration based on the
+// configuration and check used.
+func (f *filter) String() string {
+	switch f.typ {
+	case checkAnyScopes:
+		return fmt.Sprintf("%s(%s,%s)", AuthAnyScopeName, f.realm, strings.Join(f.scopes, ","))
+	case checkAllScopes:
+		return fmt.Sprintf("%s(%s,%s)", AuthAllScopeName, f.realm, strings.Join(f.scopes, ","))
+	case checkAnyKV:
+		return fmt.Sprintf("%s(%s,%s)", AuthAnyKVName, f.realm, f.kv)
+	case checkAllKV:
+		return fmt.Sprintf("%s(%s,%s)", AuthAllKVName, f.realm, f.kv)
+	}
+	return AuthUnknown
+}
+
+func (f *filter) validateRealm(h map[string]interface{}) bool {
+	if f.realm == "" {
+		return true
+	}
+
+	vI, ok := h[realmKey]
+	if !ok {
+		return false
+	}
+	v, ok := vI.(string)
+	if !ok {
+		return false
+	}
+	return v == f.realm
+}
+
+func (f *filter) validateAnyScopes(h map[string]interface{}) bool {
+	if len(f.scopes) == 0 {
+		return true
+	}
+
+	vI, ok := h[scopeKey]
+	if !ok {
+		return false
+	}
+	v, ok := vI.([]interface{})
+	if !ok {
+		return false
+	}
+	var a []string
+	for i := range v {
+		a = append(a, v[i].(string))
```

Comment:
it should be checked with `vi, ok := v[i].(string)`

Reply:
for handling if it's not a string: it even can be a good way to convert it into a string. What do you think?